### ISSUE TITLE
LEFT JOIN strange behaviour 

### ISSUE THREAD

**Issue by miyconst** [Starcounter/Starcounter#2617]
Simplified database.
Some Person, AddressRelation, Address rows in there.
Not every Person has an Address.
I want to select all Persons which has address LIKE "%moscow%".
Here is the SQL:
```SQL
SELECT p.Name, ar.Address.Name
FROM Simplified.Ring2.Person p
	LEFT JOIN Simplified.Ring3.AddressRelation ar ON ar.Somebody = p
WHERE ar.Address.Name LIKE '%moscow%'
FETCH 5
```
And here is the selection result:
![left-join-bug-1](https://cloud.githubusercontent.com/assets/6435556/6642342/00938536-c9b5-11e4-9334-d5e33d684059.PNG)
The result has some rows with NULL address which is very weird.
Another example:
```SQL
SELECT p.Name, ar.Address.Name
FROM Simplified.Ring2.Person p
	LEFT JOIN Simplified.Ring3.AddressRelation ar ON ar.Somebody = p
WHERE ar.Address.Name LIKE '%NotExistingAddress%'
FETCH 5
```
Returns:
![left-join-bug-2](https://cloud.githubusercontent.com/assets/6435556/6642388/4df09b5c-c9b5-11e4-8489-de4609a47f12.PNG)

The WHERE clause does not effect Person table, but should.
The server acts as if I would wrote such SQL:
```SQL
SELECT p.Name, ar.Address.Name
FROM Simplified.Ring2.Person p
	LEFT JOIN Simplified.Ring3.AddressRelation ar ON ar.Somebody = p AND ar.Address.Name LIKE '%NotExistingAddress%'
FETCH 5
```

**Comment by k-rus** [comment]
Thank you for your question.

> Simplified database.
> Some Person, AddressRelation, Address rows in there.
> Not every Person has an Address.
> I want to select all Persons which has address LIKE "%moscow%".
> Here is the SQL:
> 
> SELECT p.Name, ar.Address.Name
> FROM Simplified.Ring2.Person p
>     LEFT JOIN Simplified.Ring3.AddressRelation ar ON ar.Somebody = p
> WHERE ar.Address.Name LIKE '%moscow%'
> FETCH 5
> And here is the selection result:
> 
> The result has some rows with NULL address which is very weird.

Left outer join should include all persons without address by definition. To get only those, whose address satisfy the condition, inner join must be used.
Or do I understand your question correctly?

**Comment by miyconst** [comment]
If this is the default behaviour for SC then it should be documented somewhere.
In MS-SQL and MySql it works differently.

There is a search form in the People app which should return all of the Persons which has:
Name LIKE '%query%' OR Address.Name LIKE '%query%'.

For this I wrote SQL:
```SQL
SELECT p.Name, ar.Address.Name
FROM Simplified.Ring2.Person p
	LEFT JOIN Simplified.Ring3.AddressRelation ar ON ar.Somebody = p 
WHERE ar.Address.Name LIKE '%NotExistingAddress%' OR p.Name LIKE '%NotExistingAddress%'
FETCH 5
``` 

But this SQL returns all of the Persons no matter what LIKE condition is.
That is why I went dipper and found the issue with LEFT JOIN.
As you understand, I can't use INNER JOIN here, because Persons without Address will be never selected.

**Comment by k-rus** [comment]
Now, I see. You mean that the persons are returned even that the address is not null and the address does not satisfy the condition. We will look at it. I don't know if we will need your help to reproduce it. It would be great if you keep or provide a repeatable way to reproduce it.
Currently, it is expected to look into it at the end of next week.

**Comment by miyconst** [comment]
Yes, Query result contains all Persons and only those AddressRelations which satisfy WHERE condition. As you may see from the second query result, the server returns all Persons with no AddressRelations, because non of them satisfy WHERE condition.

This should happen if condition is inside JOIN clause.
WHERE clause should affect already joined rows.

I could send binaries of my db if needed.

**Comment by k-rus** [comment]
I was able to reproduce this issue.

**Comment by k-rus** [comment]
This one is difficult to fix. As I see from the implementation it was not designed to work in the expected way. We will discuss how to go with this issue.

**Comment by dan31** [comment]
We discussed subject with @k-rus. The current decision is to 1) check all platform users whether they use outer joins, 2) probably they don't, if someone does – find ways to eliminate the usage, 3) hide support for outer joins in 2.0. That support will reappear when NewSQL processor arrives.

**Comment by miyconst** [comment]
Any news for the issue?

**Comment by k-rus** [comment]
> Any news for the issue?

As @dan31 wrote, the outer join will not be supported in the current version, since as I wrote the current implementation was not designed to work correctly. Basically, our current suggestion is not to use it, but implement necessary behaviour in C#.

**Comment by miyconst** [comment]
Then should I close this issue as it won't be fixed?

**Comment by k-rus** [comment]
> Then should I close this issue as it won't be fixed?

It will be closed as you suggest after the identified tasks are completed:



> The current decision is to 1) check all platform users whether they use outer joins, 2) probably they don't, if someone does – find ways to eliminate the usage, 3) hide support for outer joins in 2.0.
